sklearn/impute/_base.py

```diff
@@ -396,17 +402,30 @@ def transform(self, X):
                 indexes = np.repeat(np.arange(len(X.indptr) - 1, dtype=np.int),
                                     np.diff(X.indptr))[mask]
 
+                sparse_constructor = (sparse.csr_matrix if X.format == 'csr'
+                                      else sparse.csc_matrix)
+                missing_mask = sparse_constructor(
+                                        (mask,
+                                         X.indices.copy(),
+                                         X.indptr.copy()),
+                                        shape=X.shape, dtype=bool)
+
                 X.data[mask] = valid_statistics[indexes].astype(X.dtype,
                                                                 copy=False)
         else:
-            mask = _get_mask(X, self.missing_values)
+            missing_mask = _get_mask(X_, self.missing_values)
+            try:
```

Comment:
Above we can define `valid_statistics_indexes` above to avoid the `try`, `except`:

```py
if self.strategy == "constant":
    valid_statistics = statistics
    valid_statistics_indexes = slice(None)
```